addon/-private/execution_context/rfc268.js

```diff
@@ -0,0 +1,194 @@
+import $ from '-jquery';
+import { resolve } from 'rsvp';
+import {
+  guardMultiple,
+  buildSelector,
+  findClosestValue,
+  getRoot
+} from '../helpers';
+import {
+  getContext,
+  visit,
+  click,
+  fillIn,
+  triggerEvent,
+  focus,
+  blur
+} from '../compatibility';
+import {
+  ELEMENT_NOT_FOUND,
+  throwBetterError
+} from '../better-errors';
+import Ceibo from 'ceibo';
+
+export default function ExecutionContext(pageObjectNode) {
+  this.pageObjectNode = pageObjectNode;
+}
+
+ExecutionContext.prototype = {
+  run(cb) {
+    return cb(this);
+  },
+
+  runAsync(cb) {
+    let root = getRoot(this.pageObjectNode);
+    let isChained = !root._chainedTree;
+
+    if (isChained) {
+      // Already chained, so our root is the root of the chained tree, and we
+      // need to wait on its promise if it has one so the previous call can
+      // resolve before we run ours.
+      root._promise = resolve(root._promise).then(() => cb(this));
+    } else {
+      // Not a chained call, so store our method's return on the chained root
+      // so that chained calls can find it to wait on it.
+      root._chainedTree._promise = cb(this);
+    }
+
+    return this.chainable();
+  },
+
+  chainable() {
+    // See explanation in `create.js` -- here instead of returning the node on
+    // which our method was invoked, we find and return our node's mirror in the
+    // chained tree so calls to it can be recognized as chained calls, and
+    // trigger the chained-call waiting behavior.
+    let root = getRoot(this.pageObjectNode);
+    let isChained = !root._chainedTree;
+
+    if (isChained) {
+      // Already chained, so our node is in the chained tree
+      return this.pageObjectNode;
+    } else {
+      // Not already chained, so we need to look up our equivalent node in the
+      // chained tree and return that. We do it by walking up the tree
+      // collecting node keys to build a path to our node, and then use that
+      // to walk back down the chained tree to our mirror node.
+      let path = [];
+      let node;
+
+      for (node = this.pageObjectNode; node; node = Ceibo.parent(node)) {
+        path.unshift(Ceibo.meta(node).key);
+      }
+      // The path will end with the root's key, 'root', so shift that back off
+      path.shift();
+
+      node = root._chainedTree;
+      path.forEach((key) => {
+        // Normally an item's key is just its property name, but collection
+        // items' keys also include their index. Collection item keys look like
+        // `foo[2]` and legacy collection item keys look like `foo(2)`.
+        let match;
+        if ((match = /\[(\d+)\]$/.exec(key))) {
+          // This is a collection item
+          let [ indexStr, index ] = match;
+          let name = key.slice(0, -indexStr.length);
+          node = node[name].objectAt(parseInt(index, 10));
+        } else if ((match = /\((\d+)\)$/.exec(key))) {
```

Comment:
Won't that make migrating pretty painful? I still have a bunch of code using legacy collections that I'm gonna migrate sometime, but haven't had time to yet since they were just released recently. The new test API has be released for quite some time and especially with Ember 3.0, many users are already using it. Do we really want to block users from using the new test API until they migrate all of their collections?

Reply:
Do you propose to release this as a part of v1 or to preserve legacy collections support in v2?
 
I think we can't do the former cause we change the way properties are defined which is a part of public API I believe. On the other hand preserving legacy collections support in v2 seems a bit strange(but acceptable at least to me). 
mhm.. Maybe it's a time for collections codemod...

Reply:
> Do you propose to release this as a part of v1 or to preserve legacy collections support in v2?

Migrating to v2 of page object is going to be a lot of work for people like @bendemboski said, I think we should strive for having this PR merged in for v1

> I think we can't do the former cause we change the way properties are defined which is a part of public API I believe

Can you expand a bit on this? I'm not sure to understand what would break.

Reply:
I'm reffering to async props like this https://github.com/san650/ember-cli-page-object/pull/357/files#diff-52672839c703d1e4a3286b560cd21215R78. We should now return an async operations. I suspect this change might break someones custom props

Reply:
Actually I'm not sure how wide-spread is custom props in the wild. 
Also seems like we haven't [documented](http://ember-cli-page-object.js.org/docs/v1.13.x/extend) declaring of async custom properties yet so it might be considered as a private API.

If that's ok I totally support including the new test API to the v1